File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.cpp
class Solution {
public:
    void solve(int cur, vector<vector<int>>&adj, vector<vector<int>>&ans, int x, vector<int>&vis) {
        vis[cur] = true;
        for(int child:adj[cur]) {
            if(!vis[child]) {
                ans[child].push_back(x);
                solve(child, adj, ans, x, vis);
            }
        }
    }
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>>ans(n), adj(n);
        for(auto e:edges) {
            adj[e[0]].push_back(e[1]);
        }
        for(int i=0; i<n; i++) {
            vector<int>vis(n, 0);
            solve(i, adj, ans, i, vis);
        }
        return ans;
    }
};

File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<List<Integer>> ans = new ArrayList<>();
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<n; i++) {
            ans.add(i, new ArrayList<>());
            adj.add(i, new ArrayList<>());
        }
        for(int[] e: edges) {
            adj.get(e[0]).add(e[1]);
        }
        for(int i=0; i<n; i++) {
            solve(i, adj, ans, i);
        }
        return ans;
    }
    public void solve(int cur, List<List<Integer>>adj, List<List<Integer>>ans, int x) {
        for(int i=0; i<adj.get(cur).size(); i++) {
            int child = adj.get(cur).get(i);
            if(ans.get(child).isEmpty() || ans.get(child).get(ans.get(child).size() - 1) != x) {
                ans.get(child).add(x);
                solve(child, adj, ans, x);
            }
        }
    }
}